### PR TITLE
filesystem: logical part numbering and related

### PR DESCRIPTION
**Pull request by dbungert**
* stricter numbering of logical partitions
* catch overallocation of primary part numbers for msdos/vtoc
* some unit test improvements

### CODE REVIEW

**Review by mwhudson — COMMENTED** [review]
Hi, this seems mostly fine, I think the server UI needs something like this though: https://github.com/dbungert/subiquity/pull/1

How will the desktop UI know when a gap is OK to have a partition in it?

**Comment by dbungert** [comment]
v2 behavior on a msdos ptable - following from the suggestions on dbungert#1 the gap is visible and you get a dialog if you attempt to use it further
image

v1 behavior on a msdos ptable - you can keep adding partitions until you run out
image
We could make this better by having the v1 code report similar unusable gaps to what the v2 code does.

I also played with having the `_can_partition_gap` return False for an unusable gap, which sounds nice on paper but is a bit of a UI mystery:
image

File: subiquity/models/filesystem.py [hunk 1]
```diff
@@ -668,6 +671,14 @@ def _decode_id(self, id):
             return None
         return id.encode('utf-8').decode('unicode_escape').strip()
 
+    def renumber_logical_partitions(self):
+        parts = [p for p in self.partitions_by_offset() if p.is_logical]
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/models/filesystem.py]: New plan then, logicals get numbered as they show up, this renumber stuff is reserved for partition deletion and only touches numbers after the deleted one.

**Review by mwhudson — APPROVED** [review]
I think this looks really good, thanks for the changes. I don't think we have the s390x bits completely straight but we can fix that after landing (some curtin work also required here).

File: subiquity/common/filesystem/gaps.py [hunk 2]
```diff
@@ -102,13 +115,26 @@ def ad(v):  # ad == "align down"
         return v - v % info.part_align
 
     def maybe_add_gap(start, end, in_extended):
+        if in_extended:
+            usable = GapUsable.YES
+        elif isinstance(device, LVM_VolGroup):
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/common/filesystem/gaps.py:120]: Fixed

File: subiquity/models/tests/test_filesystem.py [hunk 3]
```diff
@@ -676,37 +683,106 @@ def test_render_includes_unmounted_new_partition(self):
...
+    @parameterized.expand(
+        [[pt, primaries, i]
+         for pt, primaries in (('msdos', 4), ('vtoc', 3))
+         for i in range(3)]
+    )
+    def test_out_of_offset_order(self, ptable, primaries, idx_to_remove):
+        m = make_model(storage_version=2)
+        d1 = make_disk(m, ptable=ptable, size=100 << 20)
+        self.assert_next(make_partition(m, d1, size=10 << 20))
+        self.assert_next(make_partition(m, d1, flag='extended'))
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/models/tests/test_filesystem.py:751]: Saved for later.

**Pull request merged**